File: 513.find-bottom-left-tree-value.java
/*
 * @lc app=leetcode id=513 lang=java
 *
 * [513] Find Bottom Left Tree Value
 */

// @lc code=start

import java.util.ArrayDeque;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public int findBottomLeftValue(TreeNode root) {
        int ret = 0;
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            TreeNode leftMost = null;
            for (int sz = queue.size(); sz != 0; --sz) {
                TreeNode node = queue.remove();
                if (node.left != null)
                    queue.add(node.left);
                if (node.right != null)
                    queue.add(node.right);
                if (leftMost == null) {
                    leftMost = node;
                }
            }
            ret = leftMost.val;
        }
        return ret;
    }
}
// @lc code=end

File: 513.find-bottom-left-tree-value.cpp
/*
 * @lc app=leetcode id=513 lang=cpp
 *
 * [513] Find Bottom Left Tree Value
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <queue>

using namespace std;

class Solution {
public:
    int findBottomLeftValue(TreeNode* root) {
        if (root == nullptr) return -1;
        queue<TreeNode *> nodeQu;
        nodeQu.push(root);
        int ret = -1;
        while (!nodeQu.empty()) {
            bool flag = false;
            for (int sz = nodeQu.size(); sz != 0; --sz) {
                TreeNode *node = nodeQu.front(); nodeQu.pop();
                if (node->left != nullptr) nodeQu.push(node->left);
                if (node->right != nullptr) nodeQu.push(node->right);
                if (!flag) {
                    ret = node->val; flag = true;
                }
            }
        }
        return ret;
    }
};
// @lc code=end
